File: frontend/lib/models/student.dart
class Student {
  final String studentId;
  final String name;
  final String roomNo;
  final String dept;
  final String category;
  final String college;
  final bool faceRegistered;
  final String leaveStatus;
  final DateTime? leaveUntil;

  Student({
    required this.studentId,
    required this.name,
    required this.roomNo,
    required this.dept,
    required this.category,
    required this.college,
    this.faceRegistered = false,
    this.leaveStatus = 'none',
    this.leaveUntil,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      // backend may return `regNo` or `studentId` depending on server version
      studentId: json['regNo'] ?? json['studentId'] ?? '',
      name: json['name'] ?? '',
      roomNo: json['roomNo'] ?? '',
      dept: json['dept'] ?? '',
      category: json['category'] ?? '',
      college: json['college'] ?? '',
      faceRegistered: json['faceRegistered'] ?? false,
      leaveStatus: json['leaveStatus'] ?? 'none',
      leaveUntil: json['leaveUntil'] != null
          ? DateTime.tryParse(json['leaveUntil'].toString())
          : null,
    );
  }

  bool get isOnLeave => leaveStatus == 'on_leave' || leaveStatus == 'medical';
}

File: backend/src/models/Student.js
import mongoose from "mongoose";

const studentSchema = new mongoose.Schema(
  {
    name: { type: String, required: true },
    regNo: {
      type: String,
      required: true,
      unique: true,
      sparse: true,
      trim: true,
    },
    roomNo: { type: String, required: true, trim: true },
    dept: { type: String, required: true, trim: true },
    category: { type: String, required: true, trim: true },
    college: { type: String, required: true, trim: true },
    faceRegistered: { type: Boolean, default: false },
    faceEmbedding: { type: [Number], default: [] },
    createdAt: { type: Date, default: Date.now },
    leaveStatus: {
      type: String,
      enum: ["none", "on_leave", "medical"],
      default: "none",
    },
    leaveUntil: { type: Date },
  },
  { timestamps: true },
);

// Indexes for common queries
studentSchema.index({ faceRegistered: 1 });
studentSchema.index({ dept: 1 });
studentSchema.index({ roomNo: 1 });
studentSchema.index({ college: 1 });
studentSchema.index({ leaveStatus: 1 });
studentSchema.index({ createdAt: -1 });

export default mongoose.model("Student", studentSchema);
